### PR TITLE
Users can now check loaded modules for framepointers. 

### PR DESCRIPTION
**Pull request by florian-kuebler**
It is hidden behind a flag (enable_validator) for now. The result will be shown as total number of functions that do not have a valid prologue/epilogue in relation to the total number of functions.

### CODE REVIEW

File: OrbitCore/FramepointerValidatorClient.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+#include "FramepointerValidatorClient.h"
+
+#include <absl/strings/str_format.h>
+#include <absl/strings/str_join.h>
+#include <string>
+#include "CoreApp.h"
+#include "Message.h"
+#include "Pdb.h"
```
Comment (dimitry-) [inline OrbitCore/FramepointerValidatorClient.cpp:11]: clang-format?

File: OrbitCore/FramepointerValidatorClient.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      delete;
+
+  void ValidateFramepointersInModules(
+      Process* process, const std::vector<std::shared_ptr<Module>>& modules);
+
+ private:
+  void HandleResponse(const Message& message, uint64_t id);
+  TransactionClient* transaction_client_;
+  absl::flat_hash_map<uint64_t, std::vector<std::shared_ptr<Module>>>
+      id_modules_;
```
Comment (dimitry-) [inline OrbitCore/FramepointerValidatorClient.h:38]: nit: modules_map_ sounds better...

File: OrbitCore/FramepointerValidatorService.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      delete;
+  FramepointerValidatorService(FramepointerValidatorService&&) = delete;
+  FramepointerValidatorService& operator=(FramepointerValidatorService&&) =
+      delete;
+
+ private:
+  void HandleRequest(const Message& message);
+
+  static std::vector<std::shared_ptr<Function>> CheckFramepointers(
+      Pdb* pdb, bool is64Bit);
```
Comment (dimitry-) [inline OrbitCore/FramepointerValidatorService.h:36]: is_64_bit

File: OrbitCore/FramepointerValidatorService.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  std::vector<std::shared_ptr<Function>> result;
+
+  cs_mode mode = is64Bit ? CS_MODE_64 : CS_MODE_32;
+  csh handle;
+  if (cs_open(CS_ARCH_X86, mode, &handle) != CS_ERR_OK) {
+    ERROR("Unable to open capstone.");
+    return result;
+  }
+
+  cs_option(handle, CS_OPT_DETAIL, CS_OPT_ON);
```
Comment (dimitry-) [inline OrbitCore/FramepointerValidatorService.cpp:74]: can you create another class hiding all capstone-specific stuff in it... this ideally should be just call to that class..
another benefit is that you can add a unit-test for that class where service is not unit-testable.

Ideally drop dependency on Pdb... just get a file with debug symbols and pass it along.

File: OrbitCore/FramepointerValidatorService.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+  std::ifstream instream(pdb->GetFileName(), std::ios::in | std::ios::binary);
```
Comment (pierricgimmig) [inline OrbitCore/FramepointerValidatorService.cpp:76]: pdb->GetFileName() returns the name of the debug symbol file.  Unless I'm missing something, it seems you are reading the debug information file itself and not the actual module.  The bytes you send out to the validator are not code.

File: OrbitCore/CoreApp.h [hunk 6]
```diff
@@ -68,6 +69,10 @@ class CoreApp {
                        const MemoryCallback& callback) {
     process_memory_client_->GetRemoteMemory(pid, address, size, callback);
   }
+  // TODO(kuebler): hide this
```
Comment (dpallotti) [inline OrbitCore/CoreApp.h:72]: What exactly does this TODO mean? Should this be hidden now?

Also add an empty line above.

File: OrbitCore/FramepointerValidatorClient.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+  uint64_t all_functions_size = 0;
+  for (const auto& module : modules) {
+    all_functions_size += module->m_Pdb->GetFunctions().size();
+  }
+
+  std::string text = absl::StrFormat(
+      "info:Framepointer Validation\nFailed to validate %d out of %d functions",
+      functions.size(), all_functions_size);
+  GCoreApp->SendToUiNow(text);
```
Comment (dpallotti) [inline OrbitCore/FramepointerValidatorClient.cpp:71]: Pass a `CoreApp* core_app` to the constructor like in `SymbolsClient`, don't use the global directly.

File: OrbitCore/FramepointerValidatorClient.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
```
Comment (dpallotti) [inline OrbitCore/FramepointerValidatorClient.cpp:72]: Remove empty line.

File: OrbitCore/FramepointerValidatorService.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+      continue;
+    }
+
+    const bool is64Bit = process->GetIs64Bit();
+
+    // Find module.
+    const std::string& module_name = module_info.m_Name;
+    std::shared_ptr<Module> module = process->GetModuleFromName(module_name);
+    if (!module) {
+      ERROR("Unable to find module %s", module_name.c_str());
```
Comment (dpallotti) [inline OrbitCore/FramepointerValidatorService.cpp:46]: Maybe delimit the module's name. E.g.:
`ERROR("Unable to find module \"%s\"", module_name.c_str());`
or
`ERROR("Unable to find module: %s", module_name.c_str());`

Also you don't need `c_str()` as `absl::StrFormat` is used behind the scenes.

File: OrbitCore/FramepointerValidatorService.cpp [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  std::vector<ModuleDebugInfo> module_infos;
+  transaction_service_->ReceiveRequest(message, &module_infos);
+
+  for (const auto& module_info : module_infos) {
+    // Find process.
+    uint32_t pid = module_info.m_PID;
+    std::shared_ptr<Process> process = nullptr;
+    process = process_list_->GetProcess(pid);
+    if (process == nullptr) {
+      ERROR("Unable to find process %u", pid);
```
Comment (dpallotti) [inline OrbitCore/FramepointerValidatorService.cpp:36]: Maybe `"Unable to find process with pid %u"`.

File: OrbitCore/FramepointerValidatorService.cpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    const std::string& module_name = module_info.m_Name;
+    std::shared_ptr<Module> module = process->GetModuleFromName(module_name);
+    if (!module) {
+      ERROR("Unable to find module %s", module_name.c_str());
+      continue;
+    }
+
+    std::shared_ptr<Pdb> pdb = module->m_Pdb;
+    if (!pdb) {
+      ERROR("Unable to retrieve Pdb %s", module_name.c_str());
```
Comment (dpallotti) [inline OrbitCore/FramepointerValidatorService.cpp:52]: Maybe delimit the module's name.

File: OrbitGl/ModulesDataView.cpp [hunk 12]
```diff
@@ -6,7 +6,13 @@
 
 #include "App.h"
 #include "Core.h"
+#include "FramepointerValidatorService.h"
 #include "OrbitModule.h"
+#include "Pdb.h"
+#include "absl/flags/flag.h"
+
+// TODO(kuebler): remove this once we have the validator complete
+ABSL_FLAG(bool, enable_validator, false, "Enable validation of framepointers");
```
Comment (dpallotti) [inline OrbitGl/ModulesDataView.cpp:15]: Maybe `enable_frame_pointer_validator`, otherwise it's very generic?

File: OrbitGl/ModulesDataView.cpp [hunk 13]
```diff
@@ -111,12 +120,18 @@ std::vector<std::string> ModulesDataView::GetContextMenu(
       } else if (module->IsDll()) {
         enable_dll = true;
       }
+    } else {
+      enable_verify = true;
```
Comment (dpallotti) [inline OrbitGl/ModulesDataView.cpp:124]: Is there a specific reason why non-loaded modules cannot be verified?

File: OrbitGl/ModulesDataView.cpp [hunk 14]
```diff
@@ -136,23 +154,33 @@ std::vector<std::string> ModulesDataView::GetContextMenu(
 void ModulesDataView::OnContextMenu(const std::string& a_Action,
                                     int a_MenuIndex,
                                     const std::vector<int>& a_ItemIndices) {
-  if (a_Action == MENU_ACTION_MODULES_LOAD) {
+  if (a_Action == MENU_ACTION_MODULES_LOAD ||
```
Comment (dpallotti) [inline OrbitGl/ModulesDataView.cpp:157]: Can we have two separate ifs for the two actions? A few lines of code might be duplicated but it's more readable than having the logics for the two actions interleaved.

File: OrbitCore/FramepointerValidatorService.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.cpp]: Good point.

File: OrbitCore/FramepointerValidatorService.cpp [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+
+    std::vector<std::shared_ptr<Function>> functions =
+        CheckFramepointers(pdb.get(), is64Bit);
+
+    transaction_service_->SendResponse(message.GetType(), functions);
+  }
+}
+
+std::vector<std::shared_ptr<Function>>
+FramepointerValidatorService::CheckFramepointers(Pdb* pdb, bool is64Bit) {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.cpp]: I'll rename the function.

File: OrbitCore/FramepointerValidatorService.h (hunk 3 above)
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.h]: Done. Thanks.

File: OrbitCore/FramepointerValidatorService.h [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+#include <vector>
+
+#include "OrbitFunction.h"
+#include "ProcessUtils.h"
+#include "TransactionService.h"
+
+// Runs on the service and receives requests (Msg_ValidateFramepointer) from
+// FramepointerValidatorClient to validate whether certain modules are
+// compiled with framepointers.
+// It responds a list of all functions that are identified, to don't have
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.h]: Yep, this is better.

File: OrbitGl/ModulesDataView.cpp [hunk 17]
```diff
@@ -136,23 +154,33 @@ std::vector<std::string> ModulesDataView::GetContextMenu(
...
+            }
+          } else if (a_Action == MENU_ACTION_MODULES_VERIFY) {
+            modules_to_validate.push_back(module);
           }
         }
       }
     }
+    if (!modules_to_validate.empty()) {
+      GOrbitApp->GetFramepointerValidatorClient()
+          ->ValidateFramepointersInModules(m_Process.get(),
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitGl/ModulesDataView.cpp]: Sure.

File: OrbitCore/CoreApp.h (hunk 6 above)
Comment (florian-kuebler) [reply OrbitCore/CoreApp.h:72]: For the symbol client, there is a LoadRemoteModules function in `App.h`, that calls the client. So the getter of the client can be protected. 
Here, I made the frame-pointer client accessible directly.

Comment (dpallotti) [reply OrbitCore/CoreApp.h:72]: Can you add a similar explanation to the TODO comment?

Comment (florian-kuebler) [reply OrbitCore/CoreApp.h]: Sure

**Comment by florian-kuebler** [comment]
> I'll have a look on Monday, but in the meantime can we please refer to frame pointers with a space between "frame" and "pointers" (sorry German language), like in `-fomit-frame-pointer`? E.g.:
> 
> * `FramepointerValidator` -> `FramePointerValidator`,
> * `framepointer_validator` -> `frame_pointer_validator`,
> * `FRAMEPOINTER_VALIDATOR` -> `FRAME_POINTER_VALIDATOR`.

Did the renaming. Thanks.

File: OrbitCore/FramepointerValidatorService.cpp (hunk 5 above)
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.cpp]: That's a valid point here. It works, if the debug info is in the same file, which is not true for everything. 
Instead of reading in the "pdb" file, I could use the `m_FullName` of the module, or read directly in memory `(Function#Address())`. In the latter one, I would need an reinterpret_cast, from uint64_t (function address) to uint8_t*, right?
Any suggestions?

File: OrbitCore/CMakeLists.txt [hunk 18]
```diff
@@ -25,7 +25,9 @@ target_sources(
          Diff.h
          EventBuffer.h
          EventClasses.h
-         FunctionFramepointerValidator.h
+        FramePointerValidatorClient.h
```
Comment (dpallotti) [inline OrbitCore/CMakeLists.txt:28]: Indent correctly.

File: OrbitCore/CMakeLists.txt [hunk 19]
```diff
@@ -97,7 +99,9 @@ target_sources(
           Diff.cpp
           ElfFile.cpp
           EventBuffer.cpp
-          FunctionFramepointerValidator.cpp
+        FramePointerValidatorClient.cpp
```
Comment (dpallotti) [inline OrbitCore/CMakeLists.txt:102]: Indent correctly.

File: OrbitCore/CoreApp.h [hunk 20]
```diff
@@ -78,6 +85,7 @@ class CoreApp {
  private:
   std::unique_ptr<TransactionClient> transaction_client_;
   std::unique_ptr<SymbolsClient> symbols_client_;
+  std::unique_ptr<FramePointerValidatorClient> framepointer_validator_client_;
```
Comment (dpallotti) [inline OrbitCore/CoreApp.h:88]:
```suggestion
  std::unique_ptr<FramePointerValidatorClient> frame_pointer_validator_client_;
```

File: OrbitCore/CoreApp.h [hunk 21]
```diff
@@ -69,6 +70,12 @@ class CoreApp {
     process_memory_client_->GetRemoteMemory(pid, address, size, callback);
   }
 
+  // TODO(kuebler): Make this protected (as GetSymbolsClient is) and provide a
+  //  a method in the App to directly use the functionality of this client.
+  FramePointerValidatorClient* GetFramepointerValidatorClient() {
```
Comment (dpallotti) [inline OrbitCore/CoreApp.h:75]:
```suggestion
  FramePointerValidatorClient* GetFramePointerValidatorClient() {
```

File: OrbitCore/FramePointerValidatorClient.h [hunk 22]
```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_CORE_FRAMEPOINTER_VALIDATOR_CLIENT_H_
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorClient.h:5]:
```suggestion
#ifndef ORBIT_CORE_FRAME_POINTER_VALIDATOR_CLIENT_H_
```

File: OrbitCore/FramePointerValidatorClient.h [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+#define ORBIT_CORE_FRAMEPOINTER_VALIDATOR_CLIENT_H_
+
+#include "OrbitModule.h"
+#include "OrbitProcess.h"
+#include "TransactionClient.h"
+
+class CoreApp;
+
+// This class can be called from the UI on the client in order to validate
+// whether certain modules are compiled with framepointers.
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorClient.h:15]:
```suggestion
// whether certain modules are compiled with frame pointers.
```

File: OrbitCore/FramePointerValidatorClient.h [hunk 24]
```diff
@@ -0,0 +1,46 @@
...
+ private:
+  void HandleResponse(const Message& message, uint64_t id);
+  CoreApp* core_app_;
+  TransactionClient* transaction_client_;
+  absl::flat_hash_map<uint64_t, std::vector<std::shared_ptr<Module>>>
+      modules_map_;
+  absl::Mutex id_mutex_;
+};
+
+#endif  // ORBIT_CORE_FRAMEPOINTER_VALIDATOR_CLIENT_H_
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorClient.h:46]:
```suggestion
#endif  // ORBIT_CORE_FRAME_POINTER_VALIDATOR_CLIENT_H_
```

File: OrbitCore/FramePointerValidatorService.cpp [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+FramePointerValidatorService::FramePointerValidatorService(
+    const ProcessList* process_list, TransactionService* transaction_service)
+    : process_list_{process_list}, transaction_service_{transaction_service} {
+  auto on_request = [this](const Message& msg) { HandleRequest(msg); };
+  transaction_service_->RegisterTransactionRequestHandler(
+      {on_request, Msg_ValidateFramePointer, "Validate Frame Pointers"});
+}
+
+void FramePointerValidatorService::HandleRequest(const Message& message) {
+  // TODO: The code below is about 90% copy&past from SymbolsService.cpp.
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorService.cpp:26]: What should we discuss and what is there to do?

File: OrbitCore/FramePointerValidatorService.h [hunk 26]
```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_CORE_FRAMEPOINTER_VALIDATOR_SERVICE_H_
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorService.h:5]:
```suggestion
#ifndef ORBIT_CORE_FRAME_POINTER_VALIDATOR_SERVICE_H_
```

File: OrbitCore/FramePointerValidatorService.h [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+  void HandleRequest(const Message& message);
+
+  static std::vector<std::shared_ptr<Function>> GetFpoFunctions(Pdb* pdb,
+                                                                bool is_64_bit);
+
+  const ProcessList* process_list_;
+  TransactionService* transaction_service_;
+};
+
+#endif  // ORBIT_CORE_FRAMEPOINTER_VALIDATOR_SERVICE_H_
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorService.h:42]:
```suggestion
#endif  // ORBIT_CORE_FRAME_POINTER_VALIDATOR_SERVICE_H_
```

File: OrbitCore/Message.h [hunk 28]
```diff
@@ -72,6 +72,7 @@ enum MessageType : int16_t {
   Msg_DebugSymbols,
   Msg_MemoryTransfer,
   Msg_ThreadNames,
+  Msg_ValidateFramePointer,
```
Comment (dpallotti) [inline OrbitCore/Message.h:75]:
```suggestion
  Msg_ValidateFramePointers,
```

File: OrbitGl/ModulesDataView.cpp [hunk 29]
```diff
@@ -6,7 +6,14 @@
...
 #include "App.h"
 #include "Core.h"
+#include "FramePointerValidatorService.h"
 #include "OrbitModule.h"
+#include "Pdb.h"
+#include "absl/flags/flag.h"
+
+// TODO(kuebler): remove this once we have the validator complete
+ABSL_FLAG(bool, enable_frame_pointer_validator, false,
+          "Enable validation of framepointers");
```
Comment (dpallotti) [inline OrbitGl/ModulesDataView.cpp:16]:
```suggestion
          "Enable validation of frame pointers");
```

File: OrbitGl/ModulesDataView.cpp [hunk 30]
```diff
@@ -95,6 +102,8 @@ void ModulesDataView::DoSort() {
 
 //-----------------------------------------------------------------------------
 const std::string ModulesDataView::MENU_ACTION_MODULES_LOAD = "Load Symbols";
+const std::string ModulesDataView::MENU_ACTION_MODULES_VERIFY =
+    "Verify Framepointer";
```
Comment (dpallotti) [inline OrbitGl/ModulesDataView.cpp:106]:
```suggestion
    "Verify Frame Pointers";
```

File: OrbitService/OrbitAsioServer.h [hunk 31]
```diff
@@ -39,6 +40,7 @@ class OrbitAsioServer {
   std::unique_ptr<TransactionService> transaction_service_;
   std::unique_ptr<SymbolsService> symbols_service_;
   std::unique_ptr<ProcessMemoryService> process_memory_service_;
+  std::unique_ptr<FramePointerValidatorService> framepointer_validator_service_;
```
Comment (dpallotti) [inline OrbitService/OrbitAsioServer.h:43]:
```suggestion
  std::unique_ptr<FramePointerValidatorService> frame_pointer_validator_service_;
```

File: OrbitCore/FramepointerValidatorService.cpp (hunk 5 above)
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.cpp]: Decided to do the first for now (i.e. load the module instead of the pdb).

File: OrbitCore/FramePointerValidatorService.cpp (hunk 25 above)
Comment (dpallotti) [reply OrbitCore/FramePointerValidatorService.cpp:26]: Ah right, then maybe rephrase to make it clearer that you are suggesting a refactoring to share code with the other `...Service`s.

File: OrbitCore/FramepointerValidatorService.cpp (hunk 4 above)
Comment (florian-kuebler) [reply OrbitCore/FramepointerValidatorService.cpp]: Done. However, I needed to pass the list of functions instead of the pdb, as we don't want to parse the debug information twice.

File: OrbitCore/FramePointerValidator.h [hunk 32]
```diff
@@ -0,0 +1,15 @@
+#ifndef ORBIT_CORE_FRAME_POINTER_VALIDATOR_H_
+#define ORBIT_CORE_FRAME_POINTER_VALIDATOR_H_
+
+#include <vector>
+
+#include "OrbitFunction.h"
+
+class FramePointerValidator {
+ public:
+  static std::vector<std::shared_ptr<Function>> GetFpoFunctions(
```
Comment (dimitry-) [inline OrbitCore/FramePointerValidator.h:10]: Please add a unit-test

File: OrbitGl/App.cpp [hunk 33]
```diff
@@ -854,6 +854,12 @@ void OrbitApp::SendToUiNow(const std::string& message) {
   }
 }
 
+//-----------------------------------------------------------------------------
+void OrbitApp::SendInfoToUi(const std::string& title, const std::string& text) {
+  std::string message = "info:" + title + "\n" + text;
+  SendToUiNow(message);
+}
```
Comment (dimitry-) [inline OrbitGl/App.cpp:861]: looks like this should not be here - rebase went wrong?

File: OrbitGl/App.cpp [hunk 34]
```diff
@@ -854,6 +854,12 @@ void OrbitApp::SendToUiNow(const std::string& message) {
...
+
 //-----------------------------------------------------------------------------
 void OrbitApp::SendErrorToUi(const std::string& title,
                              const std::string& text) {
```
Comment (dimitry-) [inline OrbitGl/App.cpp:865]: we have it twice now.

Comment (dpallotti) [reply OrbitGl/App.cpp:865]: The previous is for an information message box, this is for an error message box.

File: OrbitGl/App.cpp (hunk 33 above)
Comment (florian-kuebler) [reply OrbitGl/App.cpp:861]: As @dpallotti said. This is for an information box, the one below is for an error box.

File: OrbitCore/FramePointerValidator.h (hunk 32 above)
Comment (florian-kuebler) [reply OrbitCore/FramePointerValidator.h:14]: I added a test. This is currently failing, as function sizes seem to be calculated incorrectly.

File: OrbitCore/FramePointerValidatorTests.cpp [hunk 35]
```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorTests.cpp:1]: The names of test files need to end with `Test.cpp`, not `Tests.cpp`. For example this allows clang-format to understand this is a test and reorder includes properly (in this case putting `#include "FramePointerValidator.h"` before other ones).

File: OrbitCore/FramePointerValidatorTests.cpp [hunk 36]
```diff
@@ -0,0 +1,51 @@
...
+
+  Pdb pdb;
+  ASSERT_TRUE(elf_file->LoadFunctions(&pdb));
+  const std::vector<std::shared_ptr<Function>>& functions = pdb.GetFunctions();
+
+  std::vector<std::shared_ptr<Function>> fpo_functions =
+      FramePointerValidator::GetFpoFunctions(functions, test_elf_file, true);
+
+  EXPECT_EQ(fpo_functions.size(), 6);
+  ExpectContainsFunctionName(fpo_functions, "_init");
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorTests.cpp:44]: Look into `EXPECT_THAT`. You can probably use something like 
`EXPECT_THAT(fpo_functions_names, ::testing::UnorderedElementsAre(...));`
and remove `ExpectContainsFunctionName`.
This allows to produce a better output when the tests fails than a simple boolean having the wrong value.

File: OrbitCore/FramePointerValidatorTests.cpp [hunk 37]
```diff
@@ -0,0 +1,51 @@
...
+
+#include "ElfFile.h"
+#include "FramePointerValidator.h"
+#include "OrbitFunction.h"
+#include "Path.h"
+
+namespace {
+
+void ExpectContainsFunctionName(
+    const std::vector<std::shared_ptr<Function>> functions, std::string name) {
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorTests.cpp:18]: const ref?

File: OrbitCore/FramePointerValidatorTests.cpp [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+#include <gtest/gtest.h>
+
+#include <vector>
+
+#include "ElfFile.h"
+#include "FramePointerValidator.h"
+#include "OrbitFunction.h"
+#include "Path.h"
+
+namespace {
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidatorTests.cpp:15]: I don't remember the reason but I think we had a discussion some time ago about the fact that there's no need to put tests inside the anonymous namespace.

File: OrbitCore/FramePointerValidatorTests.cpp (hunk 35 above)
Comment (florian-kuebler) [reply OrbitCore/FramePointerValidatorTests.cpp:1]: Okay, I was looking at the ElfFileTests.cpp... will rename this one, too.

File: OrbitCore/FramePointerValidatorTests.cpp (hunk 36 above)
Comment (florian-kuebler) [reply OrbitCore/FramePointerValidatorTests.cpp]: Thanks. Added GMock and changed the test

File: OrbitCore/FramePointerValidator.cpp [hunk 39]
```diff
@@ -0,0 +1,42 @@
...
+
+#include "FramePointerValidator.h"
+
+#include <capstone/capstone.h>
+
+#include "FunctionFramePointerValidator.h"
+#include "OrbitBase/Logging.h"
+
+std::vector<std::shared_ptr<Function>> FramePointerValidator::GetFpoFunctions(
+    std::vector<std::shared_ptr<Function>> functions, std::string file_name,
```
Comment (dpallotti) [inline OrbitCore/FramePointerValidator.cpp:13]: Can any of these be passed by const ref?

File: OrbitGl/ModulesDataView.cpp [hunk 40]
```diff
@@ -95,6 +101,8 @@ void ModulesDataView::DoSort() {
 
 //-----------------------------------------------------------------------------
 const std::string ModulesDataView::MENU_ACTION_MODULES_LOAD = "Load Symbols";
+const std::string ModulesDataView::MENU_ACTION_MODULES_VERIFY =
+    "Verify Frame Pointer";
```
Comment (dpallotti) [inline OrbitGl/ModulesDataView.cpp:105]:
```suggestion
    "Verify Frame Pointers";
```

**Comment by florian-kuebler** [comment]
I rewrote the test case using GMock (need to link against it first).

File: OrbitCore/FramePointerValidator.cpp (hunk 39 above)
Comment (florian-kuebler) [reply OrbitCore/FramePointerValidator.cpp]: Ah, yes both. Thanks for spotting this.

File: OrbitCore/FramePointerValidator.h [hunk 41]
```diff
@@ -0,0 +1,19 @@
...
```
[the remark this answers is not in the document]
Comment (dimitry-) [reply OrbitCore/FramePointerValidator.h:14]: maybe rename it to CheckFramePointerSupport

[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/FramePointerValidator.h]: I added a description. However, I am not sure whether we should change the name. I already changed it form something very similar to what you suggested to the current one, as it should make it more clear, that the result is not a boolean one.

File: OrbitCore/FramePointerValidator.cpp [hunk 42]
```diff
@@ -0,0 +1,44 @@
...
+std::vector<std::shared_ptr<Function>> FramePointerValidator::GetFpoFunctions(
+    const std::vector<std::shared_ptr<Function>>& functions,
+    const std::string& file_name, bool is_64_bit) {
+  std::vector<std::shared_ptr<Function>> result;
+
+  cs_mode mode = is_64_bit ? CS_MODE_64 : CS_MODE_32;
+  csh handle;
+  if (cs_open(CS_ARCH_X86, mode, &handle) != CS_ERR_OK) {
+    ERROR("Unable to open capstone.");
+    return result;
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply OrbitCore/FramePointerValidator.cpp]: Thanks. This makes it better. Unfortunately, I needed to convert it into a Tuble<bool, vector<...>> for serialization.
